### PR TITLE
feat: Zero-copy BufMessageReader

### PR DESCRIPTION
**Pull request by gatesn**
I suppose this should really be called "less copy"

We still copy more than we need to, but that requires a change to `VortexReadAt` to take a `bytes::BufMut` by ownership, allowing us to pass in pre-aligned buffers.

### CODE REVIEW

File: vortex-buffer/src/buffer_mut.rs [hunk 1]
```diff
@@ -411,6 +421,50 @@ impl<T> FromIterator<T> for BufferMut<T> {
...
+    unsafe fn advance_mut(&mut self, cnt: usize) {
+        if !cnt.is_multiple_of(*self.alignment) {
+            vortex_panic!(
+                "Cannot advance buffer by {} items, resulting alignment is not {}",
+                cnt,
+                self.alignment
+            );
+        }
+        self.bytes.advance(cnt);
+        self.length -= cnt;
```
Comment (gatesn) [inline vortex-buffer/src/buffer_mut.rs:460]: Hold up, this is wrong

**Review by gatesn — COMMENTED** [review]
Wait a sec

File: vortex-buffer/src/buffer_mut.rs [hunk 2]
```diff
@@ -411,6 +421,50 @@ impl<T> FromIterator<T> for BufferMut<T> {
...
+    }
+}
+
+unsafe impl BufMut for ByteBufferMut {
+    fn remaining_mut(&self) -> usize {
+        self.bytes.remaining_mut()
```
Comment (gatesn) [inline vortex-buffer/src/buffer_mut.rs]: I don't think this is correct. The bytes package changed the meaning of BufMut to allow internally resizing: https://github.com/tokio-rs/bytes/issues/170

We should probably do the same. Annoyingly, this means this function returns a huge number, e.g. `usize::MAX - self.len()`.

Comment (gatesn) [reply vortex-buffer/src/buffer_mut.rs]: Ok, fixed

File: vortex-buffer/src/buffer.rs [hunk 3]
```diff
@@ -300,6 +310,28 @@ impl From<Bytes> for ByteBuffer {
     }
 }
 
+impl Buf for ByteBuffer {
+    fn remaining(&self) -> usize {
```
Comment (lwwmanning) [inline vortex-buffer/src/buffer.rs:314]: the semantics of this trait were a bit hard to divine without checking out the code and jumping to source

**Review by lwwmanning — APPROVED** [review]

**Pull request merged**